openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -0,0 +1,453 @@
+                    unicode(self),
+                    sig_error.message,
+                )
+            )
+        return False
+
+    def _compute_digest(self, user_id):
+        """
+        Returns hash(version | session_id | user_id |)
+        """
+        hash_func = sha256()
+        for data_item in [self.version, self.session_id, user_id]:
+            hash_func.update(unicode(data_item))
+            hash_func.update('|')
+        return hash_func.hexdigest()
+
+    @staticmethod
+    def _validate_cookie_params(session_id, user_id):
+        """
+        Validates the given parameters for cookie creation.
+
+        Raises SafeCookieError if session_id is None.
+        """
+        # Compare against unicode(None) as well since the 'value'
+        # property of a cookie automatically serializes None to a
+        # string.
+        if not session_id or session_id == unicode(None):
+            # The session ID should always be valid in the cookie.
+            raise SafeCookieError(
+                "SafeCookieData not created due to invalid value for session_id '{}' for user_id '{}'.".format(
+                    session_id,
+                    user_id,
+                ))
+
+        if not user_id:
+            # The user ID is sometimes not set for
+            # 3rd party Auth and external Auth transactions
+            # as some of the session requests are made as
+            # Anonymous users.
+            log.warning(
+                "SafeCookieData received empty user_id '%s' for session_id '%s'.",
+                user_id,
+                session_id,
+            )
+
+
+class SafeSessionMiddleware(SessionMiddleware):
+    """
+    A safer middleware implementation that uses SafeCookieData instead
+    of just the session id to lookup and verify a user's session.
+    """
+    def process_request(self, request):
+        """
+        Processing the request is a multi-step process, as follows:
+
+        Step 1. The safe_cookie_data is parsed and verified from the
+        session cookie.
+
+        Step 2. The session_id is retrieved from the safe_cookie_data
+        and stored in place of the session cookie value, to be used by
+        Django's Session middleware.
+
+        Step 3. Call Django's Session Middleware to find the session
+        corresponding to the session_id and to set the session in the
+        request.
+
+        Step 4. Once the session is retrieved, verify that the user
+        bound in the safe_cookie_data matches the user attached to the
+        server's session information.
+
+        Step 5. If all is successful, the now verified user_id is stored
+        separately in the request object so it is available for another
+        final verification before sending the response (in
+        process_response).
+        """
+
+        cookie_data_string = request.COOKIES.get(settings.SESSION_COOKIE_NAME)
```

Comment:
Just a drive-by comment. Be aware that White Label/microsites will override SESSION_COOKIE_DOMAIN (but not SESSION_COOKIE_NAME) so that it can support full-hostname customization. I think this should be fine as your code is now, but I just wanted to make you aware of that.